pkg/kv/kvclient/kvcoord/dist_sender_mux_rangefeed.go

```diff
@@ -0,0 +1,243 @@
+// on an appropriate node.
+type muxer struct {
+	// eventCh receives events from all of the active muxStreams.
+	eventCh chan *roachpb.MuxRangeFeedEvent
+
+	// Context group controlling execution of MuxRangeFeed calls.
+	g ctxgroup.Group
+
+	// State pertaining to actively executing MuxRangeFeeds.
+	mu struct {
+		syncutil.Mutex
+
+		// terminalErr set when a terminal error occurs.
+		// Subsequent calls to startMuxRangeFeed will return this error.
+		terminalErr error
+
+		// Each call to start new range feed gets a unique ID which is echoed back
+		// by MuxRangeFeed rpc.
+		nextStreamID int64
+
+		// muxClient contains a nodeID->MuxRangeFeedClient.
+		muxClients map[roachpb.NodeID]roachpb.Internal_MuxRangeFeedClient
+		// activeFeeds maps nodeID to the list of streamIDs currently running on that node.
+		activeFeeds map[roachpb.NodeID][]int64
+		// producers maps streamID to the event producer -- data sent back to the
+		// consumer of range feed events.
+		producers map[int64]*channelRangeFeedEventProducer
+	}
+}
+
+func newMuxer(g ctxgroup.Group) *muxer {
+	m := &muxer{
+		eventCh: make(chan *roachpb.MuxRangeFeedEvent),
+		g:       g,
+	}
+
+	m.mu.muxClients = make(map[roachpb.NodeID]roachpb.Internal_MuxRangeFeedClient)
+	m.mu.activeFeeds = make(map[roachpb.NodeID][]int64)
+	m.mu.producers = make(map[int64]*channelRangeFeedEventProducer)
+
+	m.g.GoCtx(m.demux)
+
+	return m
+}
+
+type eventOrError struct {
+	event *roachpb.RangeFeedEvent
+	err   error
+}
+
+// channelRangeFeedEventProducer is a rangeFeedEventProducer which receives
+// events on input channel, and returns events when Recv is called.
+type channelRangeFeedEventProducer struct {
+	ctx   context.Context
+	errCh chan error // Signalled to propagate terminal error the consumer.
+	input chan eventOrError
+}
+
+// Recv implements rangeFeedEventProducer interface.
+func (c *channelRangeFeedEventProducer) Recv() (*roachpb.RangeFeedEvent, error) {
+	select {
+	case <-c.ctx.Done():
+		return nil, c.ctx.Err()
+	case err := <-c.errCh:
+		return nil, err
+	case e := <-c.input:
+		return e.event, e.err
+	}
+}
+
+var _ rangeFeedEventProducer = (*channelRangeFeedEventProducer)(nil)
+
+// startMuxRangeFeed begins the execution of rangefeed for the specified RangeFeedRequest.
+func (m *muxer) startMuxRangeFeed(
+	ctx context.Context, rpc rpcClient, req *roachpb.RangeFeedRequest,
+) (rangeFeedEventProducer, error) {
+	m.mu.Lock()
+	defer m.mu.Unlock()
+
+	if m.mu.terminalErr != nil {
+		return nil, m.mu.terminalErr
+	}
+
+	// Lookup or establish MuxRangeFeed for this node.
+	nodeID := req.Replica.NodeID
+	stream, found := m.mu.muxClients[nodeID]
+	if !found {
+		var err error
+		stream, err = rpc.InternalClient.MuxRangeFeed(rpc.ctx)
+		if err != nil {
+			return nil, err
+		}
+
+		m.mu.muxClients[req.Replica.NodeID] = stream
+		m.g.GoCtx(func(ctx context.Context) error {
+			return m.receiveEventsFromNode(ctx, nodeID, stream)
```

Comment:
Do we ever reap these goroutines when there are no further rangefeeds on the node? For example, if we spawn this goroutine here but then error on the `stream.Send()` call below, will the goroutine just keep running? Similarly if all ranges for a table move off to other nodes, will this goroutine then keep running?

Reply:
No, we don't.  I don't think it's that critical; at least not immediately... I could leave a todo.  The problem is that from the demux handler it is not clear whether the event transmitted to the consumer will cause that consumer to exit (due to e.g. split) -- and that range was the last one running on the node.  While not a hard obstacle, I didn't think it was high enough priority.  Over time, true a range can move to another node; but then... it can move back too.

Reply:
Right, I see that we never remove anything from `producers` or `activeFeeds` either. Sure, we can get this in now and fix it during stability, but I think the cleanup here needs improvement before 22.1 goes out.

Let me know if you want a preliminary approval to get this in, as long as the tests pass, and we can do another round during stability. In particular, I'd also like the stopper and tracing on the server side to be set up properly.